service/controller/chart/v1/resource/release/desired.go

```diff
@@ -0,0 +1,135 @@
+		}
+	}()
+
+	chart, err := r.helmClient.LoadChart(ctx, tarballPath)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	configMapValues, err := r.getConfigMapValues(ctx, customObject)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	secretValues, err := r.getSecretValues(ctx, customObject)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	values, err := union(configMapValues, secretValues)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	releaseState := &ReleaseState{
+		Name:    releaseName,
+		Status:  helmDeployedStatus,
+		Values:  values,
+		Version: chart.Version,
+	}
+
+	return releaseState, nil
+}
+
+func (r *Resource) getConfigMapValues(ctx context.Context, customObject v1alpha1.Chart) (map[string]interface{}, error) {
+	configMapValues := make(map[string]interface{})
+
+	if key.ConfigMapName(customObject) != "" {
+		configMapName := key.ConfigMapName(customObject)
+		configMapNamespace := key.ConfigMapNamespace(customObject)
+
+		configMap, err := r.k8sClient.CoreV1().ConfigMaps(configMapNamespace).Get(configMapName, metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			return nil, microerror.Maskf(notFoundError, "config map %#q in namespace %#q not found", configMapName, configMapNamespace)
+		} else if err != nil {
+			return nil, microerror.Mask(err)
+		}
+
+		jsonData := configMap.Data[controller.ConfigMapValuesKey]
+		if jsonData != "" {
+			err = json.Unmarshal([]byte(jsonData), &configMapValues)
+			if err != nil {
+				return nil, microerror.Mask(err)
+			}
+		}
+	}
+
+	return configMapValues, nil
+}
+
+func (r *Resource) getSecretValues(ctx context.Context, customObject v1alpha1.Chart) (map[string]interface{}, error) {
+	secretValues := make(map[string]interface{})
+
+	if key.SecretName(customObject) != "" {
+		secretName := key.SecretName(customObject)
+		secretNamespace := key.SecretNamespace(customObject)
+
+		secret, err := r.k8sClient.CoreV1().Secrets(secretNamespace).Get(secretName, metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			return nil, microerror.Maskf(notFoundError, "secret %#q in namespace %#q not found", secretName, secretNamespace)
+		} else if err != nil {
+			return nil, microerror.Mask(err)
+		}
+
+		jsonData := secret.Data[controller.SecretValuesKey]
+		if jsonData != nil {
+			err = json.Unmarshal(jsonData, &secretValues)
+			if err != nil {
+				return nil, microerror.Mask(err)
+			}
+		}
+	}
+
+	return secretValues, nil
+}
+
+func union(a, b map[string]interface{}) (map[string]interface{}, error) {
+	if a == nil {
+		return b, nil
+	}
+
+	for k, v := range b {
+		_, ok := a[k]
+		if ok {
+			// The secret and config map we use have at least one shared key. We can not
+			// decide which value is supposed to be applied.
+			return nil, microerror.Maskf(invalidConfigError, "secret and config map share the same key %s", k)
```

Comment:
I would prefer to keep the `invalidConfigError` for initializations. Anything during runtime should use something like `executionFailedError`.

Reply:
Good catch. Thx!